DATASET_VERSIONING.md

```diff
@@ -0,0 +1,76 @@
+# Dataset versioning (DRAFT)
+
+DSO API provides option to serve multiple versions of same dataset via API and Database.
+
+This document aims to cover technical details on how DSO API behaves in different scenarios, 
+depending on Dataset model configuration.
+
+
+## How dataset version is defined
+
+`Dataset` model exposes 2 parameters to control versioning:
+
+ - `Dataset.version`: string, formatted according to [Semantic Versioning standard](https://semver.org/) e.g. `0.1.0` 
+   (`<major> "." <minor> "." <patch>`).
+ - `Dataset.is_default_version`: boolean, defaults to `True`
+
+Combination of these 2 parameters defines how API and Database behave.
+
+
+## Datasets without version definition and default versions of datasets
+
+All datasets without version definition in `Dataset.version` or with `Dataset.is_default_version == True` 
+will be treated as not versioned and all tables within dataset will have API endpoint set to 
+
+> `/v1/<dataset.url_prefix>/<dataset.id>/<table.name>/`
+
+Database tables of this dataset will be prefixed with `<dataset.id>_`
+
+
+## Datasets with non-default versions
+
+Any dataset wit `version` defined and `is_default_version` is set to `False` will have API endpoint defined as:
+
+ > `/v1/<dataset.url_prefix>/<dataset.id>@<dataset.version>/<table.name>/`
+
+
+## Database structure
+
+All tables within default version of Dataset will have `<dataset.id>_` name prefix.
+
+Other Dataset versions will have `<dataset.id>_<dataset.version.major>_<dataset.version.minor>_` table name prefix.
+
+> DRAFT NOTE:
+> 
+>`<dataset.version.major>_<dataset.version.minor>` should be used with caution 
+as underscore can be lost in `to_snake_case` -> `toCamelCase` -> `to_snake_case` translation, 
+however should not come into play as table name is rather isolated.
+
+# Examples:
+
+Given following Dataset definition:
+
+ - `test` dataset with version `0.1.0` and single table `users` (fields: `id`, `name`) defined
+ - `test` dataset with version `0.1.1` is default and single table `users` (fields: `id`, `name`, `age`) defined
+ - `test` dataset with version `0.2.1` and 2 tables defined: `users` (fields: `id`, `firstName`, `lastName`) and `locations` (fields: `id`, `name`)
```

Comment:
An extra field to an existing table only increases the patch level number. But an extra table warrants an minor level number increase. Why?

Reply:
`users` table has changed, I should have highlighted that better.

Will add version changing rules to the top section.

Reply:
What i mean is that logically additions, whether being fields or tables, should be safe -> increase patch number.

All renames, type changes, deletions, are not -> increase minor (and/or major) numbers.

Or am I missing something?

Reply:
You are absolutely correct. However in example above table ‘users’ has changed structure.

Reply:
If any field gets deleted, renamed or has its type changed, that's a breaking change for consumers and SemVer calls for a major version increment.

Adding a field should be a minor increment, I think.

In this example, the major version never actually changes...

Reply:
Should it? This is not versioning example, this is database & api structure behavior. 

Document on schema versioning is going to be added to amsterdam-schema repo and outside of this document.

Reply:
Of course it's a versioning example, it's an example that does versioning :)

Reply:
Unfortunately we do not know how versioning will be working just yet. 

Major version can be reserved to dataset owner to decide if bump is need or maybe major bump comes with change of dataset source.

Versioning is up for discussion. Goal of this document is to outline what happens with database if versions change.

Reply:
Alright. I thought we did Semver, but I just saw that you wrote "formatted according to" :)